pkg/kubelet/cm/devicemanager/manager_test.go

```diff
@@ -204,9 +204,10 @@ func TestDevicePluginReRegistrationProbeMode(t *testing.T) {
 	m, ch, p1, _ := setupInProbeMode(t, devs, nil, socketName, pluginSocketName)
 
 	// Wait for the first callback to be issued.
+	// NOTE: on Windows, Unix sockets may take a bit to become active.
```

Comment:
more than 5 seconds seems like a really long time... and a 10 second timeout seems fragile. is there a way to verify the socket is active before continuing with the test, rather than relying on multi-second timeouts (comment applies in several places)?

Reply:
Currently, we are waiting for the server and the plugin sockets to become active [1], this can be seen in the logs as well:

```
I0103 13:16:55.031585    2456 stub.go:148] "Starting to serve on socket" socket="C:\\Users\\Administrator\\AppData\\Local\\Temp\\device_plugin1052575968\\device-plugin.sock"
```

And this happens before this point, so this only waits for the plugin registration callback.

But that's not the cause of the failure. As mentioned, Unix sockets may take a bit to become active. Because of this, ``IsUnixDomainSocket`` can take up to 4 seconds to return whether the given path is a unix socket or not. This is relevant because the plugin manager watcher is calling this for any new file that appeared in the ``socketDir``. We can check this by increasing the log verbosity:

```
I0103 13:16:55.051266    2456 plugin_watcher.go:160] "Handling create event" event="CREATE        \"C:\\\\Users\\\\Administrator\\\\AppData\\\\Local\\\\Temp\\\\device_plugin1052575968\\\\kubelet_internal_checkpoint\""
I0103 13:16:55.051266    2456 hostutil_windows.go:119] "Function IsUnixDomainSocket starts" filePath="C:\\Users\\Administrator\\AppData\\Local\\Temp\\device_plugin1052575968\\kubelet_internal_checkpoint"
I0103 13:16:55.051266    2456 hostutil_windows.go:126] "Dialing the socket" filePath="C:\\Users\\Administrator\\AppData\\Local\\Temp\\device_plugin1052575968\\kubelet_internal_checkpoint"
I0103 13:16:55.051266    2456 hostutil_windows.go:134] "Failed the current attempt to dial the socket, so pausing before retry" filePath="C:\\Users\\Administrator\\AppData\\Local\\Temp\\device_plugin1052575968\\kubelet_internal_checkpoint" err="dial unix C:\\Users\\Administrator\\AppData\\Local\\Temp\\device_plugin1052575968\\kubelet_internal_checkpoint: connect: No connection could be made because the target machine actively refused it." socketDialRetryPeriod="1s"
...
I0103 13:16:56.065994    2456 hostutil_windows.go:126] "Dialing the socket" filePath="C:\\Users\\Administrator\\AppData\\Local\\Temp\\device_plugin1052575968\\kubelet_internal_checkpoint"
I0103 13:16:56.065994    2456 hostutil_windows.go:134] "Failed the current attempt to dial the socket, so pausing before retry" filePath="C:\\Users\\Administrator\\AppData\\Local\\Temp\\device_plugin1052575968\\kubelet_internal_checkpoint" err="dial unix C:\\Users\\Administrator\\AppData\\Local\\Temp\\device_plugin1052575968\\kubelet_internal_checkpoint: connect: No connection could be made because the target machine actively refused it." socketDialRetryPeriod="1s"
```

So, because it takes such a long time to ensure that a file is a unix socket or not, we end up timing out. The file in this case is ``kubelet_internal_checkpoint``, which is not a unix socket, it's a file created by the device manager. This event / check is triggered multiple times, whenever the file is updated. This is why the timeout was increased here, and 10 seconds it's a safe amount.

Now, if this were to be fixed, it would be a bit more complicated.

A solution would be to separate the ``socketDir`` and the ``pluginDir`` (we can have ``pluginDir = filepath.Join(socketDir, "plugins")``. and then the plugin manager watcher will only watch ``pluginDir``, which means that the watcher will have nothing to do with the ``kublet_internal_checkpoint`` file **and** the server socket itself, which basically plagues the logs with these types of errors, trying to register to itself as a server, which doesn't make sense:

```
E0103 13:16:59.051732    2456 goroutinemap.go:150] Operation for "C:\\Users\\Administrator\\AppData\\Local\\Temp\\device_plugin1052575968\\server.sock" failed. No retries permitted until 2023-01-03 13:17:03.0517321 -0800 PST m=+8.285915701 (durationBeforeRetry 4s). Error: RegisterPlugin error -- failed to get plugin info using RPC GetInfo at socket C:\Users\Administrator\AppData\Local\Temp\device_plugin1052575968\server.sock, err: rpc error: code = Unimplemented desc = unknown service pluginregistration.Registration
```

This would mean updating the plugin registration process. Will have to check to see how deep that goes.

[1] https://github.com/kubernetes/kubernetes/blob/88c952d256b7e724a3cc0e6654680fda2c953c39/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go#L130
[2] https://github.com/kubernetes/kubernetes/blob/88c952d256b7e724a3cc0e6654680fda2c953c39/pkg/kubelet/util/util_windows.go#L140
[3] https://github.com/kubernetes/kubernetes/blob/88c952d256b7e724a3cc0e6654680fda2c953c39/pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go#L179

Reply:
resolving since this change dropped out of this PR

Reply:
This change was dropped in favor of: https://github.com/kubernetes/kubernetes/pull/115269 , which makes the changes described above.